### PR TITLE
Fix framebuffer disposal

### LINKED ISSUE #2049: Rendering freezes on iOS
Rendering sometimes freezes on iOS.  Audio continues to play and UI elements can be interacted with, but the screen will no longer update.  Difficult to consistently reproduce.

Note that the draw thread is still running, but `SwapBuffers` no longer does anything.  Potentially context/thread related, probably requires a fix in osuTK.

### PR DESCRIPTION
**Pull request by peppy**
It seems like this has always been incorrect. Causes hard-freezes on iOS but (fortunately..?) is silently handled on PC platforms.

- Closes #2049.

### CODE REVIEW

**Comment by swoolcock** [comment]
Does this resolve #2049 ?

**Comment by peppy** [comment]
It does, yes.

**Comment by smoogipoo** [comment]
I am extremely against this change.
First of all it'll break scenarios like bind_fb1 -> bind_fb2 -> bind_fb3 -> unbind -> unbind, because the last framebuffer is now static and being lost.
Secondly every framebuffer is technically bound now, meaning you can unbind arbitrary framebuffers AND the attached renderbuffers even though they're not bound.
Thirdly I don't see how this changes anything? If it's that lastframebuffer is wrong at the point of disposal's unbind then it must mean that the return of GLWrapper.BindFramebuffer is wrong and should be fixed.

**Comment by peppy** [comment]
Not sure if i'm missing something, but can you explain how this worked until now? After the first bind it would be 100% always bound, even if GlWrapper changed what FBO was bound.

I'm fine with a different fix for this, but the behaviour seems inherently incorrect in master.

**Comment by smoogipoo** [comment]
No? Look at Unbind(), it explicitly sets lastframebuffer =-1.

**Comment by smoogipoo** [comment]
I think the proper way is to maintain a stack at GLWrapper, and perform the following process:

```
FrameBuffer.Bind():
    GLWrapper.BindFramebuffer(id)

FrameBuffer.Unbind():
    GLWrapper.UnbindFramebuffer(id)

FrameBuffer.Dispose():
    GLWrapper.DisposeFramebuffer(id)

GLWrapper.BindFramebuffer(id):
    stack.Push(id)
    GL.BindFramebuffer(id)

GLWrapper.UnbindFramebuffer(id):
    if (stack.Peek() != id)
        return false
    stack.Pop()
    Bind(stack.Peek())
    return true;

GLWrapper.DisposeFramebuffer(id):
    if (!GLWrapper.UnbindFramebuffer(id))
        stack.Remove(id)
    GL.DeleteFramebuffer(id)
```

The idea is 3-fold:
1. Only the currently bound framebuffer can be unbound.
2. When unbound, the last non-disposed framebuffer is bound.
3. When a disposal happens there are two situations: either it's the last in the stack in which case we'll need to re-bind to maintain an active framebuffer, or it's somewhere else in the stack and a rebind is not needed but still needs to be removed from the stack.

**Comment by peppy** [comment]
Ah yeah, I guess we need to handle this similarly to masking etc.

Surprised it worked until now..

**Comment by peppy** [comment]
Implemented with stack, @smoogipoo please check thoroughly but don't merge until I've had a chance to test against iOS (seems to be fine on PC).

File: osu.Framework/Graphics/OpenGL/GLWrapper.cs [hunk 1]
```diff
@@ -92,7 +92,7 @@ internal static void Reset(Vector2 size)
 
             all_batches.ForEachAlive(b => b.ResetCounters());
 
-            lastFrameBuffer = DefaultFrameBuffer;
+            frame_buffer_stack.Push(DefaultFrameBuffer);
```
Comment (smoogipoo) [inline osu.Framework/Graphics/OpenGL/GLWrapper.cs]: How about clearing the stack first, as is done for everything else below?

Comment (smoogipoo) [reply osu.Framework/Graphics/OpenGL/GLWrapper.cs]: Also we want to be binding `DefaultFrameBuffer`, just in-case someone added a bind but forgot to unbind.

File: osu.Framework/Graphics/OpenGL/GLWrapper.cs [hunk 2]
```diff
@@ -526,6 +541,9 @@ internal static void DeleteFramebuffer(int frameBuffer)
         {
             if (frameBuffer == -1) return;
 
+            while (frame_buffer_stack.Peek()== frameBuffer)
```
Comment (smoogipoo) [inline osu.Framework/Graphics/OpenGL/GLWrapper.cs]:
```suggestion
            while (frame_buffer_stack.Peek() == frameBuffer)
```

**Comment by peppy** [comment]
I will be able to test properly on iOS tomorrow, so please just approve, not merge.

**Comment by peppy** [comment]
Tested to work on iOS.

**Pull request merged**